File: EvolutionaryGenerator_v2.1_Java/Generator/src/mx/tec/hermes/utils/Statistical.java
package mx.tec.hermes.utils;

/**
 * Provides a small set of statistical functions.
 * <p>
 * @author José Carlos Ortiz Bayliss (jcobayliss@gmail.com)
 * @version 1.0
 */
public abstract class Statistical {

    /**
     * Returns the mean of the values provided as argument.
     * <p>
     * @param values The values to analyze.
     * @return The mean of the values provided as argument.
     */
    public static double mean(double[] values) {
        double mean = 0;
        if (values.length == 0) {
            return 0;
        }
        for (int i = 0; i < values.length; i++) {
            mean += values[i];
        }
        return mean / values.length;
    }
       
    /**
     * Returns the standard deviation of the values provided as argument.
     * <p>
     * @param values The values to analyze.
     * @return The standard deviation of the values provided as argument.
     */
    public static double stdev(double[] values) {
        double mean, stdev;
        mean = mean(values);
        stdev = 0;
        for (int i = 0; i < values.length; i++) {
            stdev += Math.pow((values[i] - mean), 2);
        }
        if (values.length > 1) {
            return Math.sqrt(stdev / (values.length - 1));
        } else {
            return 0;
        }
    }

    /**
     * Returns the median of a values provided as argument.
     * <p>
     * @param values The values to analyze.
     * @return The median of the values provided as argument.
     */
    public static double median(double[] values) {
        double median;        
        double orderedValues[];
        /*
         * Safety check.
         */
        if (values.length == 0) {
            return Double.NaN;
        }
        orderedValues = sort(values);
        if (orderedValues.length % 2 == 1) {
            median = orderedValues[(int) (orderedValues.length / 2)];
        } else {
            median = (orderedValues[(int) (orderedValues.length / 2) - 1] + orderedValues[(int) (orderedValues.length / 2)]) / 2;
        }
        return median;
    }

    /**
     * Returns the correlation coefficient of the values provided as argument.
     * <p>
     * @param x The values to analyze.
     * @param y The values to analyze.
     * @return The correlation coefficient of the values provided as argument.
     */
    public static double correlation(double[] x, double[] y) {
        double a, b, c, d, e;
        if (x.length != y.length) {
            return Double.NaN;
        }
        a = 0;
        b = 0;
        c = 0;
        d = 0;
        e = 0;
        for (int i = 0; i < x.length; i++) {
            a += x[i] * y[i];
            b += x[i];
            c += y[i];
            d += Math.pow(x[i], 2);
            e += Math.pow(y[i], 2);
        }
        return (x.length * a - b * c) / (Math.sqrt(x.length * d - Math.pow(b, 2)) * Math.sqrt(x.length * e - Math.pow(c, 2)));
    }
    
    /**
     * Returns the lower quartile of the values provided as argument.
     * <p>
     * @param values The values to analyze.
     * @return The lower quartile of the values provided as argument.
     */
    public static double lowerQuartile(double[] values) {
        double[] set;
        /*
         * Safety check.
         */
        if (values.length == 0) {
            return Double.NaN;
        }
        set = lowerSubSet(values);
        return median(set);
    }

    /**
     * Returns the upper quartile of the values provided as argument.
     * <p>
     * @param values The values to analyze.
     * @return The upper quartile of the values provided as argument.
     */
    public static double upperQuartile(double[] values) {
        double[] set;
        /*
         * Safety check.
         */
        if (values.length == 0) {
            return Double.NaN;
        }
        set = upperSubSet(values);
        return median(set);
    }

    /**
     * Sorts the values provided as argument.
     * <p>
     * @param values The values to sort.
     * @return The values sorted in ascending order.
     */
    public static double[] sort(double[] values) {
        int i, n;
        double tempValue;
        boolean swapped;
        double orderedValues[];
        orderedValues = new double[values.length];
        for (i = 0; i < values.length; i++) {
            orderedValues[i] = values[i];
        }
        n = values.length;
        do {
            swapped = false;
            for (i = 0; i < n - 1; i++) {
                if (orderedValues[i] > orderedValues[i + 1]) {
                    tempValue = orderedValues[i + 1];
                    orderedValues[i + 1] = orderedValues[i];
                    orderedValues[i] = tempValue;
                    swapped = true;
                }
            }
            n = n - 1;
        } while (swapped);
        return orderedValues;
    }

    /**
     * Returns the maximum value in the values provided as argument.
     * <p>
     * @param values The values to analyze.
     * @return The maximum value in the values provided as argument.
     */
    public static double max(double[] values) {
        double maxValue;
        maxValue = Double.MIN_VALUE;
        /*
         * Safety check.
         */
        if (values.length == 0) {
            return Double.NaN;
        }
        for (int i = 0; i < values.length; i++) {
            if (values[i] > maxValue) {
                maxValue = values[i];
            }
        }
        return maxValue;
    }

    /**
     * Returns the minimum value in the values provided as argument.
     * <p>
     * @param values The values to analyze.
     * @return The minimum value in the values provided as argument.
     */
    public static double min(double[] values) {
        double minValue;
        minValue = Double.MAX_VALUE;
        /*
         * Safety check.
         */
        if (values.length == 0) {
            return Double.NaN;
        }
        for (int i = 0; i < values.length; i++) {
            if (values[i] < minValue) {
                minValue = values[i];
            }
        }
        return minValue;
    }

    /**
     * Returns the range of the values provided as argument.
     * <p>
     * @param values The values to analyze.
     * @return The range of the values provided as argument.
     */
    public static double range(double[] values) {
        return max(values) - min(values);
    }
    
    /**
     * Generates a set that contains only those values that are less or equal than the mean of the
     * values provided as argument.
     * <p/>
     * @param values The values to analyze.
     * @return The values that are less or equal than the mean of the values provided as argument.
     */
    private static double[] lowerSubSet(double[] values) {
        int i;
        double median;
        median = median(values);
        double subSet[], orderedValues[] = sort(values);
        subSet = new double[(int) (orderedValues.length / 2) + 1];
        for (i = 0; orderedValues[i] < median; i++) {
            subSet[i] = orderedValues[i];
        }
        subSet[i] = median;
        return subSet;
    }

    /**
     * Generates a set that contains only those values that are greater or equal than the mean of
     * the values provided as argument.
     * <p/>
     * @param values The values to analyze.
     * @return The values that are greater or equal than the mean of the values provided as
     * argument.
     */
    private static double[] upperSubSet(double[] values) {
        int i, j = 0;
        double median = median(values);
        double subSet[], orderedValues[] = sort(values);
        subSet = new double[(int) (orderedValues.length / 2) + 1];
        if (orderedValues.length % 2 == 0) {
            subSet[j++] = median;
        }
        for (i = (int) (orderedValues.length / 2); i < orderedValues.length; i++) {
            subSet[j++] = orderedValues[i];
        }
        return subSet;
    }
}

File: EvolutionaryGenerator_v2.1_Python/hermes/utils/statistical.py
"""Translation of ``mx.tec.hermes.utils.Statistical``."""

import math

from javacompat import DOUBLE_MAX_VALUE, DOUBLE_MIN_VALUE

NAN = float("nan")


def mean(values: list) -> float:
    """Returns the mean of the values provided as argument.

    :param values: The values to analyze.
    :return: The mean of the values provided as argument.
    """
    mean = 0
    if len(values) == 0:
        return 0
    for i in range(len(values)):
        mean += values[i]
    return mean / len(values)


def stdev(values: list) -> float:
    """Returns the standard deviation of the values provided as argument.

    :param values: The values to analyze.
    :return: The standard deviation of the values provided as argument.
    """
    m = mean(values)
    stdev = 0
    for i in range(len(values)):
        stdev += math.pow((values[i] - m), 2)
    if len(values) > 1:
        return math.sqrt(stdev / (len(values) - 1))
    else:
        return 0


def median(values: list) -> float:
    """Returns the median of a values provided as argument.

    :param values: The values to analyze.
    :return: The median of the values provided as argument.
    """
    #
    # Safety check.
    #
    if len(values) == 0:
        return NAN
    ordered_values = sort(values)
    if len(ordered_values) % 2 == 1:
        median = ordered_values[int(len(ordered_values) / 2)]
    else:
        median = (ordered_values[int(len(ordered_values) / 2) - 1]
                  + ordered_values[int(len(ordered_values) / 2)]) / 2
    return median


def correlation(x: list, y: list) -> float:
    """Returns the correlation coefficient of the values provided as argument.

    :param x: The values to analyze.
    :param y: The values to analyze.
    :return: The correlation coefficient of the values provided as argument.
    """
    if len(x) != len(y):
        return NAN
    a = 0
    b = 0
    c = 0
    d = 0
    e = 0
    for i in range(len(x)):
        a += x[i] * y[i]
        b += x[i]
        c += y[i]
        d += math.pow(x[i], 2)
        e += math.pow(y[i], 2)
    return ((len(x) * a - b * c)
            / (math.sqrt(len(x) * d - math.pow(b, 2)) * math.sqrt(len(x) * e - math.pow(c, 2))))


def lower_quartile(values: list) -> float:
    """Returns the lower quartile of the values provided as argument.

    :param values: The values to analyze.
    :return: The lower quartile of the values provided as argument.
    """
    #
    # Safety check.
    #
    if len(values) == 0:
        return NAN
    return median(_lower_sub_set(values))


def upper_quartile(values: list) -> float:
    """Returns the upper quartile of the values provided as argument.

    :param values: The values to analyze.
    :return: The upper quartile of the values provided as argument.
    """
    #
    # Safety check.
    #
    if len(values) == 0:
        return NAN
    return median(_upper_sub_set(values))


def sort(values: list) -> list:
    """Sorts the values provided as argument.

    :param values: The values to sort.
    :return: The values sorted in ascending order.
    """
    ordered_values = list(values)
    n = len(values)
    while True:
        swapped = False
        for i in range(n - 1):
            if ordered_values[i] > ordered_values[i + 1]:
                ordered_values[i], ordered_values[i + 1] = ordered_values[i + 1], ordered_values[i]
                swapped = True
        n = n - 1
        if not swapped:
            break
    return ordered_values


def max(values: list) -> float:
    """Returns the maximum value in the values provided as argument.

    :param values: The values to analyze.
    :return: The maximum value in the values provided as argument.

    Please note that the accumulator starts at ``Double.MIN_VALUE``, which in Java
    is the smallest *positive* double and not the most negative one.  The original
    code therefore returns 4.9E-324 when every value is negative.  The behaviour is
    preserved here on purpose so both versions produce identical results.
    """
    max_value = DOUBLE_MIN_VALUE
    #
    # Safety check.
    #
    if len(values) == 0:
        return NAN
    for i in range(len(values)):
        if values[i] > max_value:
            max_value = values[i]
    return max_value


def min(values: list) -> float:
    """Returns the minimum value in the values provided as argument.

    :param values: The values to analyze.
    :return: The minimum value in the values provided as argument.
    """
    min_value = DOUBLE_MAX_VALUE
    #
    # Safety check.
    #
    if len(values) == 0:
        return NAN
    for i in range(len(values)):
        if values[i] < min_value:
            min_value = values[i]
    return min_value


def range_(values: list) -> float:
    """Returns the range of the values provided as argument.

    :param values: The values to analyze.
    :return: The range of the values provided as argument.
    """
    return max(values) - min(values)


def _lower_sub_set(values: list) -> list:
    """Generates a set that contains only those values that are less or equal than
    the mean of the values provided as argument.

    :param values: The values to analyze.
    :return: The values that are less or equal than the mean of the values provided as argument.
    """
    median_value = median(values)
    ordered_values = sort(values)
    sub_set = [0.0] * (int(len(ordered_values) / 2) + 1)
    i = 0
    while ordered_values[i] < median_value:
        sub_set[i] = ordered_values[i]
        i += 1
    sub_set[i] = median_value
    return sub_set


def _upper_sub_set(values: list) -> list:
    """Generates a set that contains only those values that are greater or equal than
    the mean of the values provided as argument.

    :param values: The values to analyze.
    :return: The values that are greater or equal than the mean of the values provided as argument.
    """
    j = 0
    median_value = median(values)
    ordered_values = sort(values)
    sub_set = [0.0] * (int(len(ordered_values) / 2) + 1)
    if len(ordered_values) % 2 == 0:
        sub_set[j] = median_value
        j += 1
    for i in range(int(len(ordered_values) / 2), len(ordered_values)):
        sub_set[j] = ordered_values[i]
        j += 1
    return sub_set


class Statistical:
    """Provides a small set of statistical functions.

    The Java class only contains static methods, so the module level functions above
    are the actual translation.  This class is kept so that the call sites read the
    same way they do in the Java version (``Statistical.max(values)``).

    :author: Paola Azeneth Castillo Gutiérrez (Python translation)
    :original author: José Carlos Ortiz Bayliss (jcobayliss@gmail.com)
    :original version: 1.0
    """

    mean = staticmethod(mean)
    stdev = staticmethod(stdev)
    median = staticmethod(median)
    correlation = staticmethod(correlation)
    lower_quartile = staticmethod(lower_quartile)
    upper_quartile = staticmethod(upper_quartile)
    sort = staticmethod(sort)
    max = staticmethod(max)
    min = staticmethod(min)
    range = staticmethod(range_)
